pandas/tests/extension/json/array.py

```diff
@@ -104,6 +105,21 @@ def _concat_same_type(cls, to_concat):
         data = list(itertools.chain.from_iterable([x.data for x in to_concat]))
         return cls(data)
 
+    def factorize(self, na_sentinel=-1):
+        frozen = tuple(tuple(x.items()) for x in self)
+        labels, uniques = pd.factorize(frozen)
+
+        # fixup NA
```

Comment:
This type of thing is going to be error prone. We expect that `[B, NA, C, B]` is coded as `[0, -1, 1, 0]`, not `[0, 1, 2, 0]` (I ran into this with cyberpandas as well).

If library authors are using our tests, this should be caught. Otherwise, I'm not sure how to handle it. We could design a method like around `def refactorize(labels, mask, uniques)` that does something like this, but I'm not sure.

Reply:
Hmm apparently I messed this up, so the tests will fail. Fixing now.

Reply:
Fixed in 434df7dc9c927ac52b881e748ebee12c1389cf15

Reply:
If we require this, I think we should put a stronger comment (maybe as a comment below the docstring) about that we expect that missing values are not *in* the uniques.

What would go wrong if `.factorize()` on an extension array would include its missing value in the uniques?

Reply:
Hmm probably not the end of the world, but it would violate our factorize (and eventually groupby) semantics. Until recently, we messed this up for Categorical. That didn't have downstream effects for groupby, since everything in groupby is special for categorical.

I'll make this clearer.

Reply:
Well adding that simple note turned into a rabit hole. I decided to merge the 3 docstrings

- pandas.factorize
- pandas.core.base.IndexOpsMixin.factorize
- pandas.Categorical.factorize